### PR TITLE
Add “View on Perforce” button

### PR DESCRIPTION
**Pull request by vdavid**
Closes https://github.com/sourcegraph/sourcegraph/issues/39932

It reuses the same "Open on *" button that GitHub, GitLab, and Bitbucket already uses. Previously, the logic just quit in the case of Perforce. With these changes, it doesn't, but rather, it goes to a URL defined in the config.

Added a new setting to `experimentalFeatures` section of site config:
```
    "perforceCodeHostToSwarmMap": {
      "perforce1.company.com:1666": "https://swarm.company.com/"
      "perforce2.company.com:1666": "https://swarm.company.com/"
      ...
    }
```
Why in `experimentalFeatures`? Because the `perforce` setting was already there, and this doesn't make much sense without that.

The links go to `{swarmUrl}/{changelistNumber}` which gets automatically redirected to a longer URL by Swarm, so this seemed like an elegant solution (also the customer suggested this).

Caveat: I don't know Swarm's URL pattern for the file pages, so I could only link to changelists, not files. If the customer comes back to me with a preferred URL pattern, we can link directly to file pages. But this solution is only on par with what they had in their custom extension.

## Test plan

Went through all possible cases I could think of, it's in [this 4-min Loom](https://www.loom.com/share/c4c58f74f2764bf99cbe1e44876812b8)

### CODE REVIEW

File: client/shared/src/components/icons.tsx [hunk 1]
```diff
@@ -59,6 +59,19 @@ export const FormatListBulletedIcon: React.FunctionComponent<React.PropsWithChil
     </svg>
 )
 
+export const PerforceIcon: React.FunctionComponent<React.PropsWithChildren<IconProps & { color?: string }>> = props => (
```
Comment (vdavid) [inline client/shared/src/components/icons.tsx:62]: Moved this here to let it be with the rest of the code host icons. Not sure why it was separate. It was only used in one place.

File: client/web/src/extensions/components/ActionItemsBar.tsx [hunk 2]
```diff
@@ -381,7 +386,7 @@ export const ActionItemsToggle: React.FunctionComponent<React.PropsWithChildren<
             <li className={classNames('nav-item mr-2', className)}>
                 <div className={classNames(styles.toggleContainer, isOpen && styles.toggleContainerOpen)}>
                     <Tooltip content={`${isOpen ? 'Close' : 'Open'} ${panelName} panel`}>
-                        {/**
+                        {/*
```
Comment (vdavid) [inline client/web/src/extensions/components/ActionItemsBar.tsx:389]: ESLint was nagging me because of the malformatted JSDoc. But it wasn't actually a JSDoc anyway.

File: client/web/src/repo/RepoContainer.tsx [hunk 3]
```diff
@@ -353,6 +353,9 @@ export const RepoContainer: React.FunctionComponent<React.PropsWithChildren<Repo
         return null
     }
 
+    const perforceCodeHostUrlToSwarmUrlMap =
```
Comment (vdavid) [inline client/web/src/repo/RepoContainer.tsx:356]: I had to make a compromise and fetch this setting twice (both here and in `ActionItemsBar`), but I still chose this solution because this way, the `GoToCodeHostAction` module could stay decoupled from `window`. It seemed more elegant overall.

File: client/web/src/repo/RepoContainer.tsx [hunk 4]
```diff
@@ -379,19 +382,20 @@ export const RepoContainer: React.FunctionComponent<React.PropsWithChildren<Repo
                 >
                     {({ actionType }) => (
                         <GoToCodeHostAction
-                            key="go-to-code-host"
-                            source="repoHeader"
                             repo={repoOrError}
+                            repoName={repoName}
```
Comment (vdavid) [inline client/web/src/repo/RepoContainer.tsx:386]: I reordered the attributes to match their order in `GoToCodeHostAction`. They didn't change, only `perforceCodeHostUrlToSwarmUrlMap` was added.

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.03% (+0.74 kb) | 0.01% (+1.95 kb) | 0.01% (+1.20 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/dv/view-on-perforce/compare-report.html;tab=live_object) for a full comparison between the commits c87e7dc and 7b7e3a9 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

File: client/web/src/extensions/components/ActionItemsBar.tsx [hunk 5]
```diff
@@ -218,6 +218,9 @@ export const ActionItemsBar = React.memo<ActionItemsBarProps>(function ActionIte
         return <div className={styles.barCollapsed} />
     }
 
+    const perforceCodeHostUrlToSwarmUrlMap =
+        (window.context?.experimentalFeatures?.perforceCodeHostToSwarmMap as { [key: string]: string }) || {}
```
Comment (philipp-spiess) [inline client/web/src/extensions/components/ActionItemsBar.tsx:222]: I think this setting should live in the `settings` object so that orgs and users can overwrite it. My mental model is that any setting should be user configurable unless it poses a security risk. What are your thoughts here?

File: client/web/src/repo/RepoContainer.tsx (hunk 3 above)
Comment (philipp-spiess) [reply client/web/src/repo/RepoContainer.tsx:356]: The trade-of is that now `RepoContainer` and `ActionItemsBar` are coupled to `window` instead _and_ need to know about internals from `GoToCodeHostAction` which doesn't feel super nice either, wdyt?

File: client/web/src/repo/actions/GoToCodeHostAction.tsx [hunk 6]
```diff
@@ -63,6 +65,10 @@ export const GoToCodeHostAction: React.FunctionComponent<
         }, [repo, revision, filePath])
     )
 
+    const commitMessage = useObservable<string>(
+        fetchCommitMessage({ repoName: props.repoName, revision: props.revision })
```
Comment (philipp-spiess) [inline client/web/src/repo/actions/GoToCodeHostAction.tsx:69]: This looks like it might run whenever the `GoToCodeHostAction` is rendered for any code host which may cause a significant increase in network traffic.

Even if we only render this when inside a perforce repo, I assume this would add quite a lot of overhead for instances that are mainly used with perforce since it will load on every page load.

Any chance we can only load it when the button is actually clicked (or maybe start preloading when it is hovered?)

File: schema/site.schema.json [hunk 7]
```diff
@@ -626,7 +632,7 @@
               { "type": "number", "minimum": 0, "maximum": 0 },
               {
                 "type": "string",
-                "pattern": "^(unlimited|[0-9]+\\/(sec|secs|second|seconds|min|mins|minute|minutes|hr|hrs|hour|hours))$"
+                "pattern": "^(unlimited|[0-9]+/(sec|secs|second|seconds|min|mins|minute|minutes|hr|hrs|hour|hours))$"
```
Comment (philipp-spiess) [inline schema/site.schema.json:635]: This seems like an unrelated change?

File: schema/site.schema.json [hunk 8]
```diff
@@ -698,7 +704,7 @@
       "description": "Required when using any of the native code host integrations for Phabricator, GitLab, or Bitbucket Server. It is a space-separated list of allowed origins for cross-origin HTTP requests which should be the base URL for your Phabricator, GitLab, or Bitbucket Server instance.",
       "type": "string",
       "examples": ["https://my-phabricator.example.com https://my-bitbucket.example.com https://my-gitlab.example.com"],
-      "pattern": "^((https?:\\/\\/[\\w-\\.]+)( https?:\\/\\/[\\w-\\.]+)*)|\\*$",
+      "pattern": "^((https?://[\\w-.]+)( https?://[\\w-.]+)*)|\\*$",
```
Comment (philipp-spiess) [inline schema/site.schema.json:707]: Here too. What's going on? Haha I’m not sure it's equivalent and it seems a bit risky to change as part of this change.

**Review by philipp-spiess — COMMENTED** [review]
Overall this looks good to me but we should probably verify that this does not add a new network request for every page view where the go to code host action is displayed, regardless of which code host is set up.

File: schema/site.schema.json (hunk 7 above)
Comment (vdavid) [reply schema/site.schema.json:635]: Also reverted this, thanks!

**Comment by vdavid** [comment]
> @vdavid Shouldn't this feature be called "View on Swarm" 🤔

Yup, I was considering that, "View on Perforce" looked weird to me, but needed an extra push to do it. So I've got a Helix Swarm icon and added that instead of the Perforce logo. :) Will add an updated screenshot.

**Comment by vdavid** [comment]
@philipp-spiess: I've pushed an update.
- Replaced Perforce with Swarm (text and logo)
- Moved config from site config to user settings
- Now I'm sending a request for the commit message _only_ in case of Perforce, _and_ not sending a request anymore for the external links in case of Perforce, _so_ the total number of requests sent is the same as before this PR. I preferred this way because
   - the "only fetch on hover/click" could be quirky on slow/unstable connections
   - the user wouldn't immediately see the URL when hovering the button
   - the code would be more complex with little gains

[40-sec Loom](https://www.loom.com/share/875c6efb90544428aa44f3063a668d44)

WDYT?

File: client/web/src/repo/actions/GoToCodeHostAction.tsx [hunk 9]
```diff
@@ -45,22 +47,41 @@ export const GoToCodeHostAction: React.FunctionComponent<
...
+
     /**
      * The external links for the current file/dir, or undefined while loading, null while not
      * needed (because not viewing a file/dir), or an error.
      */
-    const fileExternalLinksOrError = useObservable<ExternalLinkFields[] | null | undefined | ErrorLike>(
+    const perforceCommitMessage = useObservable<string | null | undefined>(
         useMemo(() => {
-            if (!repo || !filePath) {
+            if (serviceType !== 'perforce') {
```
Comment (philipp-spiess) [inline client/web/src/repo/actions/GoToCodeHostAction.tsx:72]: We can probably also check if `Object.keys(perforceCodeHostUrlToSwarmUrlMap) > 0` as to avoid doing any additional fetching in-case no mapping is defined.

**Review by philipp-spiess — APPROVED** [review]
Amazeballs! 

Never Ending LGTM

Just one nit but I’m not sure if we can do something about it:

Screenshot 2022-10-14 at 15 44 15

The green in the icon is pretty distracting. Do you think we can change this and make it transparent instead? cc @jjinnii I’m not sure if this would change the identity of the logo too much 🤔

File: client/web/src/repo/actions/GoToCodeHostAction.tsx (hunk 9 above)
Comment (vdavid) [reply client/web/src/repo/actions/GoToCodeHostAction.tsx]: Yup, great catch, plus I've also re-added the `!repo` check as that'll prevent the button from displaying.

**Comment by twarit-waikar** [comment]
> So I've got a Helix Swarm icon and added that instead of the Perforce logo

Usually Swarm would have a favicon that's served directly by the Swarm server. You can directly embed the image from `https://swarm_url/favicon.ico`
